apps/aeutils/src/aeu_plugins.erl

```diff
@@ -36,6 +39,39 @@ check_config(PluginName, SchemaFilename, OsEnvPrefix) ->
             not_found
     end.
 
+%% Look up a user config value for a specific plugin.
+%% Key is a list of binaries, corresponding to the plugin config schema
+%% For top-level config variables, this would be a list of length 1, Example:
+%%
+%% find_config(<<"aeplugin_dev_mode">>,
+%%             [<<"microblock_interval">>],
+%%             [user_config, schema_default])
+%%
+%% Path is a list of lookup types, in priority order:
+%% * user_config - check for user-provided config data
+%% * schema_default - If a default value is configured in the config schema
+%% * {env, Application, Key} - Application environment value, if configured
+%% * {value, V} - A (constant) default value.
+%%
+find_config(PluginName, Key, Path) when is_list(Key), is_list(Path) ->
+    PluginNameBin = bin(PluginName),
+    find_config_(PluginNameBin, Key, Path).
+
+find_config_(PluginName, Key, [H|T]) ->
+    case find_config_step(PluginName, Key, H) of
+        undefined -> find_config_(PluginName, Key, T);
+        {ok, _} = Ok -> Ok
+    end;
+find_config_(_, _, []) ->
+    undefined.
+
+find_config_step(PluginName, Key, user_config) ->
+    aeu_env:find_config([<<"system">>, <<"plugins">>,
+                         {<<"name">>, PluginName, <<"config">>} | Key], [user_config]);
```

Comment:
In the path `[<<"system">>, <<"plugins">>,{<<"name">>, PluginName, <<"config">>} | Key] `am I thinking correctly that `{<<"name">>, PluginName, <<"config">>}` is essentially a compound key in a list element (plugins)? If so it seems a bit unnecessary to have the `<<"name">>` and `<<"config">>` things as part of the list key - could be clearer to have them outside, or just drop them, depending on whether there are any other element types under plugins. So something like:` [<<"system">>, <<"plugins">>,{PluginName}, <<"config">> | Key]`

(I'm thinking of this in terms of future work to regularise the schema, which is currently yang and otherwise compatible)

Reply:
The meaning of the pattern is: for `{Key1, Val1, Key2}`, find the first object where `Key1 : Val1`, then return the value corresponding to `Key2`. That is, in the following config, we want to match on `name` and traverse into `config`:

```yaml
system:
    dev_mode: true
    plugins:
        -
          name: aeplugin_dev_mode
          config:
              keyblock_interval: 0
              microblock_interval: 0
        -
```

Reply:
For further clarification, this is the internal representation of the above config (or similar):
```erlang
{<<"system">> =>
     #{<<"dev_mode">> => true,
       <<"plugins">> =>
           [#{<<"config">> =>
                  #{<<"keyblock_interval">> => 0,
                    <<"microblock_interval">> => 0},
              <<"name">> => <<"aeplugin_dev_mode">>},
            #{<<"name">> => <<"aeplugin_scm">>}]}}
```

And the normal operation of `find_config()` is that `[K1, K2 | ...]` corresponds to map keys, where the corresponding value is a new map - the subtree to traverse into.

Reply:
Your example would assume that the object is always keyed on name?
We have one other keyed object array in the schema (`mining:cuckoo:miners`), and it is keyed on "executable". Another object array (`metrics:rules`) does have a "name" attribute, but it's not a key, and not required (actually, there is no suitable key attribute, so one would presumably fetch the whole array instead).

Reply:
All good then. Understood. That works. Non keyed items could just have a implicit index in some future config setup.